mne_bids/read.py

```diff
@@ -71,6 +72,92 @@ def _read_raw(raw_fname, electrode=None, hsp=None, hpi=None, config=None,
+    # at least an event value?
+    elif 'value' in events_dict:
+        # Drop events unrelated to value
+        events_dict = _drop(events_dict, 'n/a', 'value')
+        descriptions = np.asarray(events_dict['value'], dtype=str)
+
+    # Worst case, we go with 'n/a' for all events
+    else:
+        descriptions = 'n/a'
+
+    # Add Events to raw as annotations
+    onsets = np.asarray(events_dict['onset'], dtype=float)
+    durations = np.asarray(events_dict['duration'], dtype=float)
+    annot_from_events = mne.Annotations(onset=onsets,
+                                        duration=durations,
+                                        description=descriptions,
+                                        orig_time=raw.info['meas_date'])
+    raw.set_annotations(annot_from_events)
+
+    return raw
+
+
+def _handle_channels_reading(channels_fname, raw):
+    """Read associated channels.tsv and populate raw.
+
+    Updates status (bad) and types of channels.
+    """
+    channels_dict = _from_tsv(channels_fname)
+
+    # If we have a channels.tsv file, make sure there is the optional "status"
+    # column from which to infer good and bad channels
+    if 'status' in channels_dict:
+        # find bads from channels.tsv
+        bad_bool = [True if chn == 'bad' else False
+                    for chn in channels_dict['status']]
+        bads = np.asarray(channels_dict['name'])[bad_bool]
+
+        # merge with bads already present in raw data file (if there are any)
+        unique_bads = set(raw.info['bads']).union(set(bads))
+        raw.info['bads'] = list(unique_bads)
+
+    # Furthermore, try to update the channel types according to the "type"
+    # and "name" columns in channels.tsv. These are mandatory in BIDS
+    # First, make sure that ordering of names in channels.tsv matches the
+    # ordering of names in the raw data.
+    ch_names_raw = list(raw.ch_names)
+    ch_names_json = list(channels_dict['name'])
+    if ch_names_raw == ch_names_json:
+        # Create a mapping to channel types
+        channel_type_dict = dict()
+
+        # Get the best mapping we currently have from BIDS to MNE nomenclature
+        bids_to_mne_ch_types = _get_ch_type_mapping(from_mne_to_bids=False)
+        for ch in ch_names_json:
+            # Get channel type
+            ch_type = channels_dict['type'][channels_dict['name'].index(ch)]
+
+            # Try to map from BIDS nomenclature to MNE, leave channel type
+            # untouched if we are uncertain
+            updated_ch_type = bids_to_mne_ch_types.get(ch_type, None)
+            if updated_ch_type is not None:
+                channel_type_dict[ch] = updated_ch_type
+
+        # Set the channel types in the raw data according to channels.tsv
+        raw.set_channel_types(channel_type_dict)
+    else:
+        warnings.warn('Channel names do not correspond between raw data and '
+                      'the channels.tsv file: "{}"'.format(channels_fname))
```

Comment:
raise an error, no? Can you give me a use case where they can be different?

Reply:
Yes, I'll raise an error. This is related to our conversation here: https://github.com/mne-tools/mne-bids/pull/219#discussion_r302899610